File: Yandex/Practicum/Algorightms_and_Data_Structures/7_sprint/F._Прыжки_по_лестнице/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(NM)
    Memory Complexity O(N)
    """
    MOD: int = 1000000007
    n, k = map(int, sys.stdin.readline().rstrip().split())
    dp: list[int] = [0] * (n + 1)
    dp[1] = 1
    for i in range(2, n + 1):
        for prev in range(max(1, i - k), i):
            dp[i] = (dp[i] % MOD + dp[prev] % MOD) % MOD
    sys.stdout.write(f'{dp[n]}\n')


if __name__ == '__main__':
    solution()

File: Yandex/Practicum/Algorightms_and_Data_Structures/7_sprint/F._Прыжки_по_лестнице/solution.cpp
#include <iostream>
#include <vector>


void solution() {
    /*
    Time Complexity O(NM)
    Memory Complexity O(N)
    */
    int MOD = 1000000007;
    int n, k;
    std::cin >> n >> k;
    std::vector<int> dp (n + 1, 0);
    dp[1] = 1;
    for (int i = 2; i <= n; ++i) {
        for (int prev = std::max(1, i - k); prev < i; ++prev) {
            dp[i] = (dp[i] % MOD + dp[prev] % MOD) % MOD;
        }
    }
    std::cout << dp[n] << "\n";
}


int main() {
    solution();
}